*For full screen view & to download, visit *
- `HTML`: <a href ="https://prasanth-ntu.github.io/html/ML-Course/Python-Refresher/Python-Refresher-Notes.html" target="_blank">Click here</a>
- `GITHUB`: <a href ="https://github.com/prasanth-ntu/prasanth-ntu.github.io/blob/master/html/ML-Course/Python-Refresher/Python-Refresher-Notes.ipynb" target="_blank">Click here</a>

# Introduction

<font size=4>__Topics Covered__</font>


1 [Python Basics](#Python Basics) <br>
2 [Flow Control](#Flow Control) <br>
3 [Functions](#Functions) <br>


We have to learn some basic programming concepts before doing anything in Maching Learning or Data Science or any Projects.

This notebook is created on the presumption that you have some basic grounding on programming world, with some idea on `variables & data types, arrays, flow conditions - if | else | elif | while | for | break | continue, importing libraries & functions, Functions - creating functions | passing arguments | returning values`.

<a id="Python Basics"></a>
# Python Basics

## Entering expressions

In [1]:
2+2

4

**Commonly used math operators**
- Addition 
- Subtraction
- Multiplication
- Division
- Integer Division
- Moduls/ Remainder
- Exponent/ Power

**Order of Precedence**

The order of operations (also called precedence) of Python math operators is similar to that of mathematics. The ** operator is evaluated first; the *, /, //, and % operators are evaluated next, from left to right; and the + and - operators are evaluated last (also from left to right)

In [15]:
2+3*6

20

In [16]:
(2 + 3) * 6

30

In [17]:
2**5

32

In [25]:
22/7

3.142857142857143

In [26]:
22//7

3

In [27]:
22 % 7

1

## Common data types

**Common Data Types**
- Integer - `int`
    - e.g. `0,50,-100`
- Floating-Point - `float`
    - e.g. `1.25, 3.14, -1.0, --0.5, 0.0`
- String - `str`
    - e.g. `'a', 'aa', 'aaa', 'Hello!', '11 cats'`
- Boolean - `bool`
    - e.g. `True, False`

String can be enclosed in single or double quotes

## String concatenation & Replicaiton

The meaning of an operator (e.g. `+`) may change based on the data types of the values next to it


In [23]:
# Addition for int or float
1 + 2

3

In [22]:
# Concatenation for string data type
'Bruce' + 'Wayne'

'BruceWayne'

if you try to use the + operator on a string and an integer value, Python will not know how to handle this, and it will display an <font color = red>error message</font>.

In [29]:
'Bruce' + 42

TypeError: must be str, not int

The `*` operator when used on one string value and one integer value becomes string replication operator.

In [31]:
2*5

10

In [30]:
'Batman'*5

'BatmanBatmanBatmanBatmanBatman'

**Varaibles**
- It's like box in the computer’s memory where we can store values

**Variable Names** are case sensitive.<br>We  can name a variable anything as long as it obeys the following three rules:

- It can be only one word.
- It can use only letters, numbers, and the underscore `_` character.
- It can’t begin with a number.

e.g. of *invalid variable names*: 	
`current-balance, current balance 4account, total_$um, 'batman'`

## `Hello world` - First Python program

In [4]:
# This program says hello and asks for my name.

print('Hello world!')
print('What is your name?')    # ask for their name
myName = input()
print('It is good to meet you, ' + myName)
print('The length of your name is:')
print(len(myName))
print('What is your age?')    # ask for their age
myAge = input()
print (type(myAge))
print('You will be ' + str(int(myAge) + 1) + ' in a year.')

Hello world!
What is your name?
James Bond
It is good to meet you, James Bond
The length of your name is:
10
What is your age?
50
<class 'str'>
You will be 51 in a year.


### Program Breakdown

The following line is called a comment. Python ignores comments, and you can use them to write notes or remind yourself what the code is trying to do. Any text for the rest of the line following a hash mark `#` is part of a comment.

#### Comments

In [5]:
# This program says hello and asks for my name.

The `print()` function displays the `string` value inside the parentheses on the screen. Notice that the quotes are not printed to the screen. They just mark where the string begins and ends; they are not part of the string value.

#### `print()` function

In [6]:
print('Hello world!')
print('What is your name?')    # ask for their name

Hello world!
What is your name?


The `input()` function waits for the user to type some text on the keyboard and press ENTER key.
The user input will then be stored in `myName` variable

#### `input()` function

In [7]:
myName = input()

Gandolf the Gray


The expression `'It is good to meet you, ' + myName` is evaluated and is `print`ed on the screen.

In [8]:
print('It is good to meet you, ' + myName)

It is good to meet you, Gandolf the Gray


The `len()` function returns the number of characters in the given string.

#### `len()` function

In [9]:
print('The length of your name is:')
print(len(myName))

The length of your name is:
16


#### type() function

The `type()` function returns the data type of the value given

In [7]:
print('What is your age?')    # ask for their age
myAge = input()
print (type(myAge))           # the input functions usually returns a string data type

What is your age?
50
<class 'str'>


#### `str(), int(), float()` functions

These functions are used to change the data types of the values. <br>
As the `input()` function returns `str` data, `myAge` variable is of `str` data type. <br>
To add a numerical value to it, the value should be converted to `int`.<br> 
After performing the addition, the `int` value is converted back to `str` again so that it can be concatenated with other strings.

In [8]:
print('You will be ' + str(int(myAge) + 1) + ' in a year.')

50


<a id="Flow Control"></a>
# Flow Control

## `and,or,not` and `<,>,<=,>=,==,!=`

- Boolean Values: `True, False`
- Boolean operators: `and,or,not`
    - returns `True` or `False`
- Comparison operators: `<, >, <=, >=, ==, !=`
    - returns `True` or `False`

In [14]:
5 > 6

(False, True, False)

In [20]:
5 <= 6, 3>5, 3!=5, 3 == 5

(True, False, True, False)

In [17]:
5<6 and 3>5, 5<6 or 3>5

(False, True)

In [12]:
not (True), not(False)

(False, True)

## Flow control

### `if,else,elif`

#### `if`

- When `if` condition evalues to `True`, the block of statement indented inside will get executed. 
- The block can be indented (pushed in) by *`Tab`*. The block can be deindented by *`Shift+Tab`*

*Note: There should be atleast one line of indented code after `if` condition*

In [30]:
age = 5
if age < 10:                       # age < 10 evaluates to True      
    print ('age is less than 10')  # if True, this code will get executed

age is less than 10


In [31]:
age = 15
if age < 10:                       # age < 10 evaluates to False             
    print ('age is less than 10')  # Doesn't get printed as age variable us more than 10

#### `else`

- When `if` condition evalues to `True`, the block of statement indented inside `if` will get executed.
- When it evaluates to `False`, the  block of statement indented inside `else` will get executed.

In [32]:
age = 18
if age < 18:                     
    print ('You are not an adult')
else:
    print ('You are an adult')

You are an adult


#### `elif`

- When `if` condition evalues to `True`, the block of statement indented inside `if` will get executed.
- When it fails, the `elif` condition will be evaluated and if it turns out to be `True`,  the block of statement indented inside `elif` will get executed.
- When both `if` and `elif` evaluates to `False`, the  block of statement indented inside `else` will get executed.

In [37]:
age = 10
if age > 18:
    print ('You are  an adult')
elif age > 12:
    print ('You are a teenager')
else:
    print ('You are neither adult nor teen. Just a kid!')

You are neither adult nor teen. Just a kid!


We can have multiple `elif` statements after an `if` statement. In the entire combo, `if-elseif-else` combo, only one of the condition will be exeucted. The order of execution is from top to bottom.

In [41]:
age = 15
if age >60 :
    print ('You are very old')
elif age>18:
    print ('You are an adult')
elif age > 12:
    print ('You are a teenager')
else:
    print ('You are neither adult nor teen. Just a kid!')

You are a teenager


#### nested `if`

We can have `if-else` inside another `if-else` and we can keep nesting them. The deeper the nest gets, the harder it will be to debug and trace.

In [44]:
user_name = input('Enter user name: ')      # james bond
pass_word = input('Enter pass word: ')      # 007
if user_name == 'james bond':
    if pass_word == '007':
        print ('Access granted')
    else:
        print ('Access denied. Wrong password')
else:
     print ('Access denied. Wrong username')

Enter user name: james bond
Enter pass word: 008
Access denied. Wrong password


The above code could be re-witten as follows using `boolean` and `comparison operators` making it easy to read and follow

In [46]:
user_name = input('Enter user name: ')      # james bond
pass_word = input('Enter pass word: ')      # 007
if user_name == 'james bond' and pass_word == '007':
    print ('Access granted')
else:
    print ('Access denied. Wrong login details')

Enter user name: james bond
Enter pass word: 007
Access granted


### Loops

#### `while` loop

- `while` loop keeps on running as long as the condition evaluates to `True`

In [1]:
i = 10
print ('Loop start')
while i>0:            # while loop evaluates to true as long as i>0
    print (i)
    i = i - 1         # everytime, the loop runs, value of `i` decreases by 1
print ('Loop end')    # This line of code is outside the loop. Will run only when the while loop ends

Loop start
10
9
8
7
6
5
4
3
2
1
Loop end


What happens if you run **`while False:`** and **`while True:`**?

#### for loop

- `for` loop runs for a fixed number of iterations.
- Here, the `range(10)` returns a list of numbers from 0 to 9.
- Everytime, the `for` loop runs, the iterator variable `i` takes the next available value from the list.

In [3]:
for i in range(10):
    print (i)

0
1
2
3
4
5
6
7
8
9


- Beside using `range()` function in `for` loop, we can use many other data types.
- Here, I have used `list` type containing list of 4 names and used `for` loop to read through all names in list, one at a time .

*Note: We will see more about list data types later.*

In [5]:
for name in ['Batman', 'Superman', 'Ironman', 'Thanos']:
    print (name)

Batman
Superman
Ironman
Thanos


- We can also loop through each element in a `string` data type

*Note: We will see more about string data types later.*

In [6]:
for letter in 'Bruce Wayne':
    print (letter)

B
r
u
c
e
 
W
a
y
n
e


#### `break`

`break` is used to exit the loop (`while` or `for`) the program is currently evaluating prematurely based on certain conditions.

In [9]:
print ('Loop start')
for letter in 'Bruce Wayne':
    print (letter)
    if letter == 'e':
        print ('Break the loop')
        break                     # this command breaks the for loop (takes the program outside the loop)
print ('Loop end')

Loop start
B
r
u
c
e
Break the loop
Loop end


#### `continue`

- `continue` is used to skip the execution of code inside the loop (`for` or `while`) and move on to next iteration on certain conditions.

The program below skips printing the vowel letters and prints the rest of the consonants. 

In [11]:
print ('Loop start')
for letter in 'Bruce Wayne':
    if letter == 'a' or letter =='e' or letter =='i' or letter =='o' or letter =='u':
        continue      # this command skips lines inside the for loop below this line and starts the next iteration
    print (letter)
               
print ('Loop end')

Loop start
B
r
c
 
W
y
n
Loop end


### `in`

- The `in` operator can be used to check whether a value is inside another and it evaluates to `bool` (`True`,`False`)

*Note: We will see more about `in`, `str` & `list` data types later.*

In [14]:
'a' in 'James'

True

In [20]:
'b' in 'James', 'Ja' in 'James', 'James' in 'James', 'JameS' in 'James'

(False, True, True, False)

In [13]:
print ('Loop start')
for letter in 'Bruce Wayne':
    if letter in 'aeiou':          
        continue      # this command skips lines inside the for loop below this line and starts the next iteration
    print (letter)
               
print ('Loop end')

Loop start
B
r
c
 
W
y
n
Loop end


<a id="Functions"></a>
# Functions

## `import`

**built-in functions**
- Till now, we have used a basic set of functions like `print(), inpu()t, len(), str(), int(), float()`.

**standard library**
- Set of modules that conatins related group of functions embedded together.
- e.g.
    - `math` module contains mathematic related functions.
    - `random` module containsrandom number related functions.       

> Functions are often used as *"Black Boxes"*. Often, all we need to know is its inputs (parameter values), outputs and its generic function. We don't have to burden to understand each and every function and the code inside it.

**`random`** library

In [24]:
import random                # Before using a function in standard library, we have to use import _______
print (random.randint(1,10)) # Randomly generates an integer from 1 to 10

5


In [23]:
print (random.randint(1,10))  # Randomly generates an integer from 1 to 10

1


**`math`** library

In [30]:
import math
math.sqrt(2)

1.4142135623730951

In [31]:
math.pi

3.141592653589793

## Function - `def`

- A `function` is like mini program inside a program.
- The idea behind function is to avoid duplicating code.
- Deduplicaiton makes our code shorter, easier to read and update.

### fn - without `arguments` and `return`

In [47]:
# create a function named 'welcome'. It neither take any input and nor returns any output
def welcome():
    print ('hi')
    print ('mi casa es tu casa')

In [46]:
# invoking/ calling a function
welcome()

hi
mi casa es tu casa


In [43]:
welcome()
welcome()

hi
mi casa es tu casa
hi
mi casa es tu casa


### fn - with `arguments`

In [51]:
def hello(name):
    print ('Hello ' + name + '!')
    print ('Bienvenido a casa')

In [52]:
hello('Scare Crow')

Hello Scare Crow!
Bienvenido a casa


### fn - with `return`

- e.g.
    - `len()` - returns the lenght of the argument passed inside
    - `len('Python')` >>> 5

In [54]:
def greet(lan):
    if lan == 'english':
        return 'Welcome'
    elif lan == 'spanish':
        return 'Bienvenido'
    elif lan == 'french':
        return 'Bonjour'
    else:
        return 'Welcome'

In [56]:
greet('english')

'Welcome'

In [59]:
greet('french')

'Bonjour'

## Local and Global scope

- *Local scope* - Variables inside the function exist in local scope
    - Created whenever a function is called
    - Variables created inside a function eist within local scope and they cannot be used outside the function
    - When the function returns/ ends, the scope is closed/ destroyed along with the variables inside it
    - The next time we call the function, the local scope is created again and they wont remember the previous values 
    
- *Global scope* - Variables that are assigned outside all functions exist in global scope
    - Created when our program begins
    - When program ends, the global scope is destroyed

A variable must be either in local or global scope, cannot be both. 

*Note: We will see more about `Local and Global scope` later.*